core/internal/filestream/filestream.go

```diff
@@ -59,8 +52,13 @@ type FileStream interface {
 		offsetMap FileStreamOffsetMap,
 	)
 
-	// Close waits for all work to be completed.
-	Close()
+	// FinishWithExit marks the run as complete and blocks until all
+	// uploads finish.
+	FinishWithExit(exitCode int32)
+
+	// FinishWithoutExit blocks until all uploads finish but does not
+	// mark the run as complete.
+	FinishWithoutExit()
```

Comment:
does it mean that in this case the backend will keep assuming that this run is still running? when do we get into this case?

Reply:
See the usage in `sender.go`: we invoke this if there is no exit record. It shouldn't happen: as far as I know `sendExit` is the only place we create a Defer request, and it sets `exitRecord`. But that's not local to this piece of code, so it's a sketchy assumption to make.

I considered calling `FinishWithExit` with an explicit exit code here, but it muddles the semantics of the exit code---you would no longer be able to assume it comes from the user's script.

Reply:
Added `CaptureError` for this case!